compiler/rustc_trait_selection/src/solve/assembly.rs

```diff
@@ -90,16 +90,20 @@ pub(super) trait GoalKind<'tcx>: TypeFoldable<'tcx> + Copy + Eq {
 
     fn trait_def_id(self, tcx: TyCtxt<'tcx>) -> DefId;
 
-    fn consider_impl_candidate(
+    // Use a hypothesis, which consists of a "assumption" and some "requirements",
+    // to satisfy a goal. If the requirements hold, then attempt to satisfy our
+    // goal by equating it with the assumption.
+    fn consider_hypothesis(
```

Comment:
not too happy with hypothesis https://www.merriam-webster.com/dictionary/hypothesis

> A [hypothesis](https://www.merriam-webster.com/dictionary/hypothesis) is an assumption, an idea that is proposed for the sake of argument so that it can be tested to see if it might be true.
> 
> In the scientific method, the hypothesis is constructed before any applicable research has been done, apart from a basic background review. You ask a question, read up on what has been studied before, and then form a hypothesis.

we know that the `requirements -> assumption` is correct. Looking at chalk, they use `clause` for this. I don't think we should do that in rustc because we already have a `Clause` which (at least for now) is something slightly different.

Other ideas:
- `consider_(logical)_statement` (meh)
- `consider_implication` (this will be how implications in the param env will be implemented (if and once we get them))
- `consider_implied_clause` (changing `assumption` to `clause: ty::Clause`, will actually re)

I personally kinda like `consider_implied_clause` :thinking: It is a clause which is implied by some `requirements`. But yeah, if  you don't like `consider_implied_clause` we can just go with whatever.

r=me after nit

Reply:
Yeah, consider_implied_clause is a totally fine name!